src/core/queryClient.ts

```diff
@@ -129,14 +130,28 @@ export class QueryClient {
 
   setQueryData<TData>(
     queryKey: QueryKey,
-    updater: Updater<TData | undefined, TData>,
+    updater: Updater<TData | undefined, TData> | undefined,
     options?: SetDataOptions
-  ): TData {
-    const parsedOptions = parseQueryArgs(queryKey)
-    const defaultedOptions = this.defaultQueryOptions(parsedOptions)
-    return this.queryCache
-      .build(this, defaultedOptions)
-      .setData(updater, options)
+  ): TData | undefined {
+    if (typeof updater === 'undefined') {
+      return undefined
+    }
+
+    let query = this.queryCache.find<TData>(queryKey)
+    const prevData = query?.state.data
+    const data = functionalUpdate(updater, prevData)
+
+    if (typeof data === 'undefined') {
+      return
+    }
+
+    if (!query) {
+      const parsedOptions = parseQueryArgs(queryKey)
+      const defaultedOptions = this.defaultQueryOptions(parsedOptions)
+      query = this.queryCache.build(this, defaultedOptions)
+    }
+
+    return query.setData(data, options)
```

Comment:
if we go this way, I think there would be no need for `query.setData` to accept a functional updater, as we now only call it with `data`. That would allow us to get rid of some code in `query.ts` that we have re-introduced here.

However, from our discussion on the other PR, I thought we were going a slightly different route by _disallowing_ `undefined` as a cache value altogether.

It is a common source of error that people do not return anything from their `queryFn` because they do things like:

```
axios.get('/api').then(result => console.log(result))
```

which then returns `Promise<void>`. Of course, TypeScript helps :)

But I like the current approach much more because it's simpler. So how about we do:

- the proposed change here to `query.setData`
- add a dev mode error logging here if we get `undefined` just to make people aware that this is disallowed
https://github.com/tannerlinsley/react-query/blob/c0fc916cb3e7d32a05e9356b2f1359c3cd5de573/src/core/query.ts#L446
- try to tweak the typescript types to disallow `undefined` for `TData` on type level - if that's possible. `null` should still be allowed though.
- we'd need to update the docs (migration guide), but I can do that too if you want

Reply:
@TkDodo If I understand you correctly, the changes that I made so far is good. That is, if `setQueryData` is passed undefined data or a function that returns undefined, bail out. We just need additional changes to prevent undefined being written to the cache. Per #2925:

> Also, if a fetch function returns undefined, we should:
> * transform it to a failed promise (error)
> * warn during development mode about it

That is, I need to modify `setData` as you propose above (including removing extra functional update handling.)

I will also look into the type change.

I will look at the docs and see if I can update them satisfactorily.

Reply:
Looking at your comments again, I believe you are proposing to _not_ fail a query when the queryFn returns undefined, but instead to bail out. This is what I tried before, and I discovered it breaks clients that are expecting the query status to change as a result of a fetch. I recommend we stick with your original proposal and fail the query if the queryFn returns undefined.

Reply:
Transforming to error is fine for me 👍